src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -1735,5 +1759,63 @@ private RendezvousAwaitable<int> InternalReceiveResponseHeadersAsync(WinHttpRequ
 
             return state.LifecycleAwaitable;
         }
+
+        internal void ChangeCleanerTimer(TimeSpan timeout)
+        {
+            Debug.Assert(_certificateCleanupTimer != null);
+            if (_certificateCleanupTimer!.Change(timeout, Timeout.InfiniteTimeSpan))
+            {
+                isTimerRunning = timeout != Timeout.InfiniteTimeSpan;
+            }
+        }
+
+        private void ClearStaleCertificates()
+        {
+            Debug.Assert(_certificateCleanupTimer != null);
+
+            foreach (KeyValuePair<IPAddress, CachedCertificateValue> kvPair in cachedCertificates)
+            {
+                if (IsStale(kvPair.Value.LastUsedTime))
+                {
+                    cachedCertificates.TryRemove(kvPair.Key, out CachedCertificateValue _);
+                }
+            }
+
+            ChangeCleanerTimer(cachedCertificates.IsEmpty ? Timeout.InfiniteTimeSpan : s_cleanCachedCertificateTimeout);
+
+            static bool IsStale(long lastUsedTime)
+            {
+                long now = Stopwatch.GetTimestamp();
+                return (now - lastUsedTime) > s_staleTimeout;
+            }
+        }
+
+        internal void CheckTimer()
+        {
+            if (cachedCertificates.IsEmpty && isTimerRunning)
+            {
+                ChangeCleanerTimer(Timeout.InfiniteTimeSpan);
+            }
+
+            if (!cachedCertificates.IsEmpty && !isTimerRunning)
+            {
+                ChangeCleanerTimer(s_cleanCachedCertificateTimeout);
+            }
+        }
+
+        internal bool GetCertificateFromCache(IPAddress ipAddress, [MaybeNullWhen(false)] out byte[] rawCertificateBytes)
+        {
+            if (cachedCertificates.TryGetValue(ipAddress, out CachedCertificateValue cachedCert))
+            {
+                rawCertificateBytes = cachedCert.RawCertificateData;
+                cachedCert.LastUsedTime = Stopwatch.GetTimestamp();
```

Comment:
`CachedCertificateValue` is a struct, so this assignment is currently a no-op as it's modifying a local copy. IMO the cleanest option is to make `CachedCertificateValue` a class.